doc/adr/0011-concept-exception-handling-backend.md

```diff
@@ -0,0 +1,33 @@
+# 10. How to use exceptions in the backend
+
+Date: 2024-06-28
+
+## Status
+
+Accepted
+
+## Context
+So far we used different approaches when to use Optionals, throw checked exceptions or unchecked exceptions. This also led
+to using exceptions from other layers. We want to streamline our approach.
+
+## Decision
+Layer-Specific Exceptions
+* By having each layer (ports, adapters, and application) define its own exceptions, we achieve a clear separation of concerns. This ensures that exceptions are relevant to the specific context in which they occur.
+* Avoiding a common exception package prevents mixing unrelated exceptions and promotes a more organized codebase.
+
+Custom Runtime Exceptions
+* Throwing custom runtime exceptions in the application layer allows us to encapsulate domain-specific logic. These exceptions can be tailored to the norms project’s requirements.
+* Using @ControllerAdvice to handle these exceptions ensures consistent error handling across the application. It centralizes exception handling logic and provides a single point of control.
+
+Output Ports
+* Returning optionals from output ports aligns with functional programming principles. Optionals allow for more elegant handling of absence or optional results.
+* Translating exceptions thrown in output ports at the application layer maintains a clean boundary between layers. It avoids leaking implementation details and promotes a clear contract.
+* The output port is not allowed not have a dependency on the application layer’s exceptions.
+
+Input Ports
+* Allowing input ports to depend on application layer exceptions acknowledges that input ports may need to handle specific application-related errors.
```

Comment:
And this is by default right? We don't need to allow it extra but since the input ports are implemented in the application layer of course they can access application exceptions

Reply:
The thought here was that we define the rules in our architecture tests. We could have decided otherwise.